x/exchange/market_test.go

```diff
@@ -0,0 +1,4198 @@
+			toRemove: []sdk.Coin{coin(1, "finger")},
+			expErr:   "",
+		},
+		{
+			name:     "one one different second amount",
+			toAdd:    []sdk.Coin{coin(1, "finger")},
+			toRemove: []sdk.Coin{coin(2, "finger")},
+			expErr:   "",
+		},
+		{
+			name:     "one one different second denom",
+			toAdd:    []sdk.Coin{coin(1, "finger")},
+			toRemove: []sdk.Coin{coin(1, "toe")},
+			expErr:   "",
+		},
+		{
+			name:     "three three two common",
+			field:    "body part",
+			toAdd:    []sdk.Coin{coin(1, "finger"), coin(2, "toe"), coin(3, "elbow")},
+			toRemove: []sdk.Coin{coin(5, "toe"), coin(3, "elbow"), coin(1, "finger")},
+			expErr:   "cannot add and remove the same body part options 1finger,3elbow",
+		},
+		{
+			name:     "adding one invalid",
+			field:    "badabadabada",
+			toAdd:    []sdk.Coin{coin(0, "zerocoin")},
+			toRemove: nil,
+			expErr:   `invalid badabadabada to add option "0zerocoin": amount cannot be zero`,
+		},
+		{
+			name:     "adding two invalid entries",
+			field:    "friendly iterative error lookup device (F.I.E.L.D)",
+			toAdd:    []sdk.Coin{coin(-1, "bananas"), coin(3, "okaycoin"), coin(0, "zerocoin"), coin(5, "goodgood")},
+			toRemove: nil,
+			expErr: joinErrs(
+				`invalid friendly iterative error lookup device (F.I.E.L.D) to add option "-1bananas": negative coin amount: -1`,
+				`invalid friendly iterative error lookup device (F.I.E.L.D) to add option "0zerocoin": amount cannot be zero`,
+			),
+		},
+		{
+			name:     "removing one invalid",
+			field:    "",
+			toAdd:    nil,
+			toRemove: []sdk.Coin{coin(-1, "bananas")},
+			expErr:   "",
+		},
+		{
+			name:     "adding and removing one invalid",
+			field:    "fruits",
+			toAdd:    []sdk.Coin{coin(-1, "bananas")},
+			toRemove: []sdk.Coin{coin(-1, "bananas")},
+			expErr: joinErrs(
+				`invalid fruits to add option "-1bananas": negative coin amount: -1`,
+				"cannot add and remove the same fruits options -1bananas",
+			),
+		},
+		{
+			name:  "multiple errors",
+			field: "merrs!",
+			toAdd: []sdk.Coin{
+				coin(1, "apple"),
+				coin(3, "l"),
+				coin(99, "banana"),
+				coin(-55, "peach"),
+				coin(14, "copycoin"),
+				coin(5, "grape"),
+			},
+			toRemove: []sdk.Coin{
+				coin(98, "banana"),
+				coin(14, "copycoin"),
+			},
+			expErr: joinErrs(
+				`invalid merrs! to add option "3l": invalid denom: l`,
+				`invalid merrs! to add option "-55peach": negative coin amount: -55`,
+				"cannot add and remove the same merrs! options 14copycoin",
+			),
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var err error
+			testFunc := func() {
+				err = ValidateAddRemoveFeeOptions(tc.field, tc.toAdd, tc.toRemove)
+			}
+			require.NotPanics(t, testFunc, "ValidateAddRemoveFeeOptions")
+
+			assertions.AssertErrorValue(t, err, tc.expErr, "ValidateAddRemoveFeeOptions")
+		})
+	}
+}
+
+func TestValidateAddRemoveFeeOptionsWithExisting(t *testing.T) {
+	coin := func(amount int64, denom string) sdk.Coin {
```

Comment:
sdk.NewInt64Coin.

Reply:
* `sdk.NewInt64Coin` is 20 characters long, `coin` is 4. So a) the important bits are a larger percent of the text, and b) one or two extra calls can be put on a single line before it gets long enough to be split up.
* `sdk.NewInt64Coin` runs validation and panics if the denom doesn't match the regex or the amount is negative. That might be handy for production code, but it's often detrimental to unit tests.
* `sdk.NewInt64Coin` has the arguments backwards. Whenever anyone talks or thinks about coins, it's `<amount>` `<denom>`, but for some dumb reason, the SDK decided they wanted the denom first. So this `coin` function is easier to read.